### PR TITLE
Add "Set up environment" module

### PR DESCRIPTION
**Pull request by jaguarfi**
# Description

A new module is created to guide the reader through the installation of the required software and the setting up of the development environment.

Create this new module required to move some of the existing chapters to different parts of the manual, including the new module.

# Testing

The content was tested manually. The test was merely to check that the content of the manual was rendered as intended. 

# Is it [Done](https://wiki.aalto.fi/display/EDIT/Definition+of+Done)?

*Clean up your git commit history before submitting the pull request!*

- [X] I (developer) have tested the changes manually
- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -8,12 +8,16 @@ Other content is not visible in A+.
   :maxdepth: 2
```
Comment (markkuriekkinen) [inline index.rst:8]: This is a good time to change the name of the course at the top of this file: from "A sample course" to "A+ manual".

File: introduction/gallery.rst [hunk 2]
```diff
@@ -92,7 +94,7 @@ Essay grading
 
 .. figure:: /images/gallery/rubyric.png
 
-   `Rubyric is a tool for grading essays <../m06_rubyric/01_introduction>`_.
+   `Rubyric is a tool for grading essays <../rubyric/introduction>`_.
```
Comment (markkuriekkinen) [inline introduction/gallery.rst]: "Rubyric is a tool for grading essays": this is actually a little misleading since Rubyric is above all a tool for manual grading using rubrics. Manual grading is used more widely than just in essays, for example, large programming projects that are too open-ended for automatic grading.

File: introduction/index.rst [hunk 3]
```diff
@@ -1,16 +1,8 @@
...
 
-  01_installation
-  0X_gallery
-  02_rst
-  03_css
-  04_git
-  05_docker
-  06_languages
-  07_questionnaires
-  enrollment_en
```
Comment (markkuriekkinen) [inline introduction/index.rst:16]: Where did you move the enrollment questionnaire (`enrollment_en`)? It is not a normal chapter, hence it could still be better to keep it under the first module, but where is it now?

File: languages/index.rst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Languages
+=========
+
+Introduction to developing course on A-Plus learning management service.
```
Comment (markkuriekkinen) [inline languages/index.rst]: A+ does not display this text written in the module index anywhere, but nonetheless, this sentence does not describe this module at all. If this module is about multilingual courses, you could write that here.

File: set_up_environment/docker.rst [hunk 5]
```diff
@@ -1,6 +1,3 @@
-Docker
```
Comment (markkuriekkinen) [inline set_up_environment/docker.rst:1]: I think there should be a title in the beginning of the chapter. Why did you remove this completely?

File: set_up_environment/docker.rst [hunk 6]
```diff
@@ -102,6 +99,9 @@ develop a course. This is *required* if you plan to include your course
 exercises that are more than simple quizzes.
 
 .. image:: /images/aplus-architecture.png
+  :align: center
+
+|
```
Comment (markkuriekkinen) [inline set_up_environment/docker.rst:102]: What does this single vertical bar do here?

File: set_up_environment/editors.rst [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+Text editors and IDEs
+---------------------
+
+The course content, as mentioned in previous chapters, consist of several **RST** files. Therefore, you are going to need 
+a versatile and powerful :abbr:`text editor (A text editor is a type of computer program that edits plain text. Such
+programs are sometimes known as "notepad" software.)`. We suggest using text editors with advanced functionalities,
+such as Vim, or Emacs. Nevertheless, we advocate the use of :abbr:`IDEs (An integrated development environment (IDE) is a
+software application that provides comprehensive facilities to computer programmers for software development. An IDE
+normally consists of at least a source code editor, build automation tools and a debugger.)` over any text editors, 
+especially VS Code and the `A+ Tools extension <https://marketplace.visualstudio.com/items?itemName=jaguarfi.aplus-tools>`_.
```
Comment (markkuriekkinen) [inline set_up_environment/editors.rst:36]: Now that I see the URL of the VS Code extension, I noticed that you have used your personal account to publish it. Should we create an organization called apluslms so that the extension could be published under the organization? It would then be more natural for other team members to update the extension, especially if you leave Aalto.

File: set_up_environment/editors.rst [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+Some more advanced and cross-platform editors are:
+
+* `Atom <https://atom.io/>`_
+* `VS Code <https://code.visualstudio.com/>`_ (Recommended by the A+ team)
+
+.. warning::
+
+  * Do **not** use word processors such as Microsoft Word or LibreOffice writer. Their formatting commands are basically
+    useless when one must edit plain text files.
+  * Some of the most advanced editors might have a steep learning curve, but once you have mastered those editor the
```
Comment (markkuriekkinen) [inline set_up_environment/editors.rst]: The latter part of the sentence has issues. 
How about "but once you have mastered the editor's features, your productivity will increase"?

File: set_up_environment/first_steps.rst [hunk 9]
```diff
@@ -0,0 +1,209 @@
+First steps
+===========
+
+.. styled-topic::
+
+  Main questions:
+      How to run a template course? and how to install the required software in your computer?
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: remove the first question mark in the middle of the sentence

File: set_up_environment/first_steps.rst [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+  The above image list the content of the `Aplus course template <https://github.com/apluslms/aplus-course-template>`_
+  directory, in a terminal window.
+
+.. admonition:: Important: UNIX Shell, Terminal emulator
+  :class: alert alert-warning
+
+  The terminal is used for moving and copying files, starting software
+  and examining their error messages. To be precise, there is the terminal
+  emulator window and inside it another program, `shell`, which reads
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: RST uses two backticks in inline monospace text.

```
``shell``
```

File: set_up_environment/first_steps.rst [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+  make --version
+
+You can also consult the official documentation for
+`installing git <https://git-scm.com/book/en/v2/Getting-Started-Installing-Git>`_, and also for
+`installing make <https://pkgs.org/download/build-essential>`_.
+
+.. warning::
+  if you are using an Aalto's computer, you will need administrator rights to install the required software; otherwise,
+  you can ask an administrator to install them for you.
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: In Aalto Linux, you can install packages without sudo: `aptdcon --install git`. It works with packages available in the default repositories, so docker can't be installed that way.

File: set_up_environment/first_steps.rst [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+The first step to start authoring a course is to install some required software on your computer. The first two software
+you need to install are `git <https://git-scm.com/>`_ and `make <https://www.gnu.org/software/make/>`_. Sometimes, these
+two software are installed by default, but if **git** and **make** are not installed by default, you can install them by
+running the :ref:`commands below <install-git>` in your computer.
+
+.. code-block:: bash
+  :name: install-git
+  :caption: 1- Install Git
+
+  sudo apt update
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: I think it is more typical to use `apt-get` and not `apt` here. 
`sudo apt-get update`, `sudo apt-get install git` and so on for all of these code blocks.

File: set_up_environment/first_steps.rst [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+`installing make <https://pkgs.org/download/build-essential>`_.
+
+.. warning::
+  if you are using an Aalto's computer, you will need administrator rights to install the required software; otherwise,
+  you can ask an administrator to install them for you.
+  You can contact `Aalto IT services <https://www.aalto.fi/en/service-entities/it-services>`_ for requesting
+  administrator rights for the Aalto's computers.
+
+All other software runs inside `Docker <https://www.docker.com/>`_ containers. For short, Docker is the new software
+platform for running A+ and software related to it. This course will discuss Docker later.
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: "Docker is the new software platform": this is vague. You could say that A+ and related services are run in containers so that you don't need to install them manually in your computer for testing and developing courses.

File: set_up_environment/first_steps.rst [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+  emulator window and inside it another program, `shell`, which reads
+  user commands and acts according to them. If you are new to this, read the
+  `"Learning the shell" LinuxCommand.org <http://linuxcommand.org/lc3_learning_the_shell.php>`_
+  tutorial parts 1-6.
+
+Software prerequisites
+----------------------
+
+The first step to start authoring a course is to install some required software on your computer. The first two software
+you need to install are `git <https://git-scm.com/>`_ and `make <https://www.gnu.org/software/make/>`_. Sometimes, these
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: git is really required, but `make` is not required if the course RST code is compiled with our Docker container. But I suppose the VS Code extension requires `make` in order to compile and render RST inside the editor.

File: set_up_environment/first_steps.rst [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+ information about the most convenient text editor to create *RST* documents
+ :ref:`at the end of this module <text-editor-and-ides>`.
+
+Cloning this codebase with git
+------------------------------
+
+A good way to start is to clone the `Aplus course Template`_ with
+git and begin to work on it. In Aalto, you could use
+`version.aalto.fi repo <https://version.aalto.fi/gitlab/course/aplus-manual>`_. Please note that if you don't have
+access, then you can request access to the *Aalto IT services*. In that case, you probably should be part of
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: Teachers request access to https://version.aalto.fi/gitlab/course/aplus-manual from the EDIT team, not from Aalto ITS. But they can access the manual in Github so they really don't need access in version.aalto.fi.
This paragraph also mixes the new course template with the manual. You should clarify the wording here. The manual is used to learn about the system and the course template is used as a starting point for developing new courses.

File: set_up_environment/first_steps.rst [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+apluslms-cs@aalto.fi email list too.
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: It is not right to introduce our Aalto-internal email lists (apluslms-cs) as a side thought when the paragraph discusses a completely different topic. Nonetheless, the email list need not be covered here. We have an internal wiki for that internal contact information.

File: set_up_environment/first_steps.rst [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+
+To clone, execute the following command:
+
+.. code-block:: sh
+
+    git clone git@version.aalto.fi:course/aplus-manual.git
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: Do we need to mention the private version.aalto.fi repo at all? Everyone can clone the public Github repo. We should rename the Github repo from course-templates to aplus-manual. I could do it even now before merging this PR.

File: set_up_environment/first_steps.rst [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+.. code-block:: sh
+
+    ./docker-compile.sh
+
+This command runs Sphinx inside a Docker container. Sphinx reads all RST files.
+It produces compilation results into new directory ``_build``. That contains
+three subdirectories. ``doctrees`` has intermediate results of compilation.
+``html`` has the HTML versions of the RST documentation. ``yaml`` contains
+configuration files in `YAML format <https://en.wikipedia.org/wiki/YAML>`_
+and they are also meant for A+.
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: "meant for A+": better wording "they are also used by the MOOC-Grader with A+"

File: set_up_environment/first_steps.rst [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+For now, you can test the result by running A+ on your own computer:
+
+.. code-block:: sh
+
+    ./docker-up.sh
+
+Now A+ runs at http://localhost:8000. Open that address with your web browser.
+You must login by the maintenance login showing on the front page. The available
+users are `root`:`root` and `student`:`student`.  The default course is created
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: The teacher and assistant users should be listed here too.

File: set_up_environment/first_steps.rst [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+    running on your machine.
+6. Examine the changes you made in A+.
+7. Press :kbd:`Q` or :kbd:`Ctrl` ``+`` :kbd:`C` in the terminal to quit and remove data
+    or :kbd:`S` or :kbd:`ESC` to quit and keep data.
+8. Go to step 2 if you wish to continue editing.
+9. Give command ``git add -u`` in the terminal to mark all changed files to be
+    added into your local git repository.
+10. Give command ``git commit -m "message"`` in terminal. Replace ``message`` with a short description (preferably less
+than 60 characters) on what you have done.
+11. Remember also to say ``git push`` a couple of times in a day. This copies your updates to the course material to the
```
Comment (markkuriekkinen) [inline set_up_environment/first_steps.rst]: `git push` here assumes that the repo was cloned from a course repo instead of the aplus-manual repo. Previously, readers were instructed to clone the manual repo, but they are not supposed to push their own course contents there.

File: style_aplus/css.rst [hunk 21]
```diff
@@ -82,11 +82,11 @@ Contents of the CSS file
 ------------------------
 
 After compiling the course, open the files
-``_build/html/m01_introduction/02_rst.html`` and ``_static/course.css``
+``_build/html/introduction/rst.html`` and ``_static/course.css``
 in a text editor.
 
 Here is the original ``admonition`` block from file
-``m01_introduction/02_rst.rst`` :
+``introduction/rst.rst`` :
```
Comment (markkuriekkinen) [inline style_aplus/css.rst]: space before the colon `:`?

File: set_up_environment/vs_code.rst [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+the `download <https://code.visualstudio.com/download>`_ section of the official website.
+
+
+.. _vs-extensions:
+
+Install the recommended VS Code extensions
+------------------------------------------
+
+The `Aplus Tools extension <https://marketplace.visualstudio.com/items?itemName=jaguarfi.aplus-tools>`_ created for VS
+Code does provide a set of functionalities that will reduce the time needed for creating course content, some of the
```
Comment (markkuriekkinen) [inline set_up_environment/vs_code.rst]: does provide -> provides

Comment (markkuriekkinen) [reply set_up_environment/vs_code.rst]: Replace the comma with a period and start a new sentence: Some of the most relevant features are listed below:

File: set_up_environment/vs_code.rst [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+  .. figure:: /images/gallery/snippets.png
+    :width: 60 %
+    :align: center
+  
+    Small list of snippets available in the Aplus Tools extension.
+
+In order to install the **Aplus tools extension** you should visit the
+`VS Code Marketplace <https://marketplace.visualstudio.com/items?itemName=jaguarfi.aplus-tools#getting-starteds>`_ and
+follow the instructions presented in the getting started section. If you experience any problem during the
+installation or setup of your extension please contact aplusguru@aalto.fi
```
Comment (markkuriekkinen) [inline set_up_environment/vs_code.rst]: The primary email domain is cs.aalto.fi, but an alias for aalto.fi was also created.

File: set_up_environment/vs_code.rst [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+.. note::
+
+  You can develop the course content without using VS Code. However, we recommend to set up the VS Code environment 
+  because this will help you to reduce the strain of writing RST code. 
+
```
Comment (markkuriekkinen) [inline set_up_environment/vs_code.rst]: Extra empty lines in the end could be removed.

**Review by markkuriekkinen — CHANGES_REQUESTED** [review]
Comments below. 

I can't write comments to the images directly so this one is here.
**images/gallery/syntax-highlighting.png**: the screenshot shows a different kind of _templates/layout.html file than we really use. A+ courses should extend "aplus/layout.html" and use the block extrahead for adding custom CSS and JS files.

File: introduction/index.rst (hunk 3 above)
Comment (jaguarfi) [reply introduction/index.rst:16]: It looks like I removed it accidentally.

File: set_up_environment/docker.rst (hunk 6 above)
Comment (jaguarfi) [reply set_up_environment/docker.rst:102]: Since the image does not have a proper margin at the bottom, I had to add whitespace manually.

File: set_up_environment/editors.rst (hunk 7 above)
Comment (jaguarfi) [reply set_up_environment/editors.rst:36]: Yes, I was worried about that. We need to create an account for aplus.

File: set_up_environment/first_steps.rst (hunk 14 above)
Comment (jaguarfi) [reply set_up_environment/first_steps.rst]: I took this from the already existing **first_steps.rst** chapter. https://plus.cs.aalto.fi/aplus-manual/master/introduction/installation/#software-prerequisites. Should I remove make from the list of prerequisites?

File: set_up_environment/first_steps.rst (hunk 15 above)
Comment (jaguarfi) [reply set_up_environment/first_steps.rst]: Yes, you are right. I will not mention https://version.aalto.fi here. GitHub is the right place to start. Regarding the wording, my idea here is to start using the [new course template](https://github.com/apluslms/aplus-course-template) so that we will not refer to the course manual any more. But first, we need to make the new course template official.

File: set_up_environment/first_steps.rst (hunk 17 above)
Comment (jaguarfi) [reply set_up_environment/first_steps.rst]: https://github.com/apluslms/course-templates/pull/28/files#r504636795

File: set_up_environment/first_steps.rst (hunk 14 above)
Comment (markkuriekkinen) [reply set_up_environment/first_steps.rst]: Yes, make can be removed. But does VS Code require it or can VS Code compile RST with the Docker container?

Comment (jaguarfi) [reply set_up_environment/first_steps.rst]: The extension is not able to compile the course using the docker container. That is something that I will have to check in the future.

**Comment by jaguarfi** [comment]
I had to force-push. I rebased the branch into master, but I have applied all the required changes.

**Review by markkuriekkinen — APPROVED** [review]
Excellent! The VS Code extension link still contains the personal account. Has it been published with the A+ LMS account yet?

**Comment by jaguarfi** [comment]
Yes, the extension will keep the publisher, otherwise, the extension will stop working for those who have already downloaded it. But, the new publisher "apluslms" should be able to completely manage the extension. Any case, I sent another email asking about changing the URL, but I have not gotten a response.
